// @daostack/arc.js/dist/lib/plugins/utils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("../index");
exports.ProposalPlugins = {
    FundingRequest: index_1.FundingRequest,
    JoinAndQuit: index_1.JoinAndQuit,
    GenericScheme: index_1.GenericPlugin,
    SchemeRegistrar: index_1.SchemeRegistrarPlugin,
    ContributionReward: index_1.ContributionRewardPlugin,
    ContributionRewardExt: index_1.ContributionRewardExtPlugin,
    Competition: index_1.CompetitionPlugin
};
exports.Plugins = __assign(__assign({}, exports.ProposalPlugins), { ReputationFromToken: index_1.ReputationFromTokenPlugin, unknown: index_1.UnknownPlugin });
exports.Proposals = {
    GenericScheme: index_1.GenericPluginProposal,
    ContributionReward: index_1.ContributionRewardProposal,
    Competition: index_1.CompetitionProposal,
    ContributionRewardExt: index_1.ContributionRewardExtProposal,
    FundingRequest: index_1.FundingRequestProposal,
    JoinAndQuit: index_1.JoinAndQuitProposal,
    SchemeRegistrar: index_1.SchemeRegistrarProposal,
    SchemeRegistrarAdd: index_1.SchemeRegistrarProposal,
    SchemeRegistrarEdit: index_1.SchemeRegistrarProposal,
    SchemeRegistrarRemove: index_1.SchemeRegistrarProposal
};
