// @daostack/arc.js/dist/lib/genesisProtocol.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bn_js_1 = __importDefault(require("bn.js"));
var index_1 = require("./index");
function mapGenesisProtocolParams(params) {
    return {
        activationTime: Number(params.activationTime),
        boostedVotePeriodLimit: Number(params.boostedVotePeriodLimit),
        daoBountyConst: Number(params.daoBountyConst),
        limitExponentValue: Number(params.limitExponentValue),
        minimumDaoBounty: new bn_js_1.default(params.minimumDaoBounty),
        preBoostedVotePeriodLimit: Number(params.preBoostedVotePeriodLimit),
        proposingRepReward: new bn_js_1.default(params.proposingRepReward),
        queuedVotePeriodLimit: Number(params.queuedVotePeriodLimit),
        queuedVoteRequiredPercentage: Number(params.queuedVoteRequiredPercentage),
        quietEndingPeriod: Number(params.quietEndingPeriod),
        thresholdConst: index_1.realMathToNumber(new bn_js_1.default(params.thresholdConst)),
        votersReputationLossRatio: Number(params.votersReputationLossRatio)
    };
}
exports.mapGenesisProtocolParams = mapGenesisProtocolParams;
