// @daostack/arc.js/dist/lib/reward.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bn_js_1 = __importDefault(require("bn.js"));
var graphql_tag_1 = __importDefault(require("graphql-tag"));
var index_1 = require("./index");
var Reward = /** @class */ (function (_super) {
    __extends(Reward, _super);
    function Reward() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Reward.search = function (context, options, apolloQueryOptions) {
        if (options === void 0) { options = {}; }
        if (apolloQueryOptions === void 0) { apolloQueryOptions = {}; }
        var _a, _b;
        var where = '';
        if (!options.where) {
            options.where = {};
        }
        var itemMap = function (arc, item, queriedId) {
            var state = Reward.itemMap(arc, item, queriedId);
            return new Reward(arc, state);
        };
        var proposalId = options.where.proposal;
        // if we are searching for stakes on a specific proposal (a common case), we
        // will structure the query so that stakes are stored in the cache together wit the proposal
        if (proposalId) {
            delete options.where.proposal;
        }
        for (var _i = 0, _c = Object.keys(options.where); _i < _c.length; _i++) {
            var key = _c[_i];
            if (options.where[key] === undefined) {
                continue;
            }
            if (key === 'beneficiary' || key === 'dao') {
                var option = options.where[key];
                index_1.isAddress(option);
                options.where[key] = option.toLowerCase();
            }
            where += key + ": \"" + options.where[key] + "\"\n";
        }
        var query;
        if (proposalId) {
            query = graphql_tag_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["query RewardSearchFromProposal\n      {\n        proposal (id: \"", "\") {\n          id\n          gpRewards ", " {\n            ...RewardFields\n          }\n        }\n      }\n      ", "\n      "], ["query RewardSearchFromProposal\n      {\n        proposal (id: \"", "\") {\n          id\n          gpRewards ", " {\n            ...RewardFields\n          }\n        }\n      }\n      ", "\n      "])), proposalId, index_1.createGraphQlQuery(options, where), Reward.fragments.RewardFields);
            return context.getObservableObject(context, query, function (arc, r, queriedId) {
                if (!r) {
                    return [];
                }
                var rewards = r.gpRewards;
                var itemMapper = function (item) {
                    var state = Reward.itemMap(arc, item, queriedId);
                    return new Reward(arc, state);
                };
                return rewards.map(itemMapper);
            }, (_a = options.where) === null || _a === void 0 ? void 0 : _a.id, apolloQueryOptions);
        }
        else {
            query = graphql_tag_1.default(templateObject_2 || (templateObject_2 = __makeTemplateObject(["query RewardSearch\n      {\n        gprewards ", " {\n          ...RewardFields\n        }\n      }\n      ", "\n      "], ["query RewardSearch\n      {\n        gprewards ", " {\n          ...RewardFields\n        }\n      }\n      ", "\n      "])), index_1.createGraphQlQuery(options, where), Reward.fragments.RewardFields);
        }
        return context.getObservableList(context, query, itemMap, (_b = options.where) === null || _b === void 0 ? void 0 : _b.id, apolloQueryOptions);
    };
    Reward.itemMap = function (context, item, queriedId) {
        if (!item) {
            throw Error("Reward ItemMap failed. " + (queriedId && "Could not find Reward with id '" + queriedId + "'"));
        }
        return {
            beneficiary: item.beneficiary,
            createdAt: item.createdAt,
            daoBountyForStaker: new bn_js_1.default(item.daoBountyForStaker),
            daoBountyForStakerRedeemedAt: Number(item.daoBountyForStakerRedeemedAt),
            id: item.id,
            proposalId: item.proposal.id,
            reputationForProposer: new bn_js_1.default(item.reputationForProposer),
            reputationForProposerRedeemedAt: Number(item.reputationForProposerRedeemedAt),
            reputationForVoter: new bn_js_1.default(item.reputationForVoter),
            reputationForVoterRedeemedAt: Number(item.reputationForVoterRedeemedAt),
            tokenAddress: item.tokenAddress,
            tokensForStaker: new bn_js_1.default(item.tokensForStaker),
            tokensForStakerRedeemedAt: Number(item.tokensForStakerRedeemedAt)
        };
    };
    Reward.prototype.state = function (apolloQueryOptions) {
        if (apolloQueryOptions === void 0) { apolloQueryOptions = {}; }
        var query = graphql_tag_1.default(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n      query RewardState {\n        gpreward (id: \"", "\")\n        {\n          ...RewardFields\n        }\n      }\n      ", "\n    "], ["\n      query RewardState {\n        gpreward (id: \"", "\")\n        {\n          ...RewardFields\n        }\n      }\n      ", "\n    "])), this.id, Reward.fragments.RewardFields);
        return this.context.getObservableObject(this.context, query, Reward.itemMap, this.id, apolloQueryOptions);
    };
    Reward.fragments = {
        RewardFields: graphql_tag_1.default(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n      fragment RewardFields on GPReward {\n        id\n        createdAt\n        dao {\n          id\n        }\n        beneficiary\n        daoBountyForStaker\n        proposal {\n          id\n        }\n        reputationForVoter\n        reputationForVoterRedeemedAt\n        reputationForProposer\n        reputationForProposerRedeemedAt\n        tokenAddress\n        tokensForStaker\n        tokensForStakerRedeemedAt\n        daoBountyForStakerRedeemedAt\n      }\n    "], ["\n      fragment RewardFields on GPReward {\n        id\n        createdAt\n        dao {\n          id\n        }\n        beneficiary\n        daoBountyForStaker\n        proposal {\n          id\n        }\n        reputationForVoter\n        reputationForVoterRedeemedAt\n        reputationForProposer\n        reputationForProposerRedeemedAt\n        tokenAddress\n        tokensForStaker\n        tokensForStakerRedeemedAt\n        daoBountyForStakerRedeemedAt\n      }\n    "])))
    };
    return Reward;
}(index_1.Entity));
exports.Reward = Reward;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
