// @daostack/arc.js/dist/lib/stake.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bn_js_1 = __importDefault(require("bn.js"));
var graphql_tag_1 = __importDefault(require("graphql-tag"));
var index_1 = require("./index");
var Stake = /** @class */ (function (_super) {
    __extends(Stake, _super);
    function Stake() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Stake.search = function (context, options, apolloQueryOptions) {
        if (options === void 0) { options = {}; }
        if (apolloQueryOptions === void 0) { apolloQueryOptions = {}; }
        var _a, _b;
        if (!options.where) {
            options.where = {};
        }
        var where = '';
        var itemMap = function (arc, item, queriedId) {
            var state = Stake.itemMap(arc, item, queriedId);
            return new Stake(arc, state);
        };
        var proposalId = options.where.proposal;
        // if we are searching for stakes on a specific proposal (a common case), we
        // will structure the query so that stakes are stored in the cache together wit the proposal
        if (proposalId) {
            delete options.where.proposal;
        }
        for (var _i = 0, _c = Object.keys(options.where); _i < _c.length; _i++) {
            var key = _c[_i];
            if (options.where[key] === undefined) {
                continue;
            }
            if (key === 'staker' || key === 'dao') {
                var option = options.where[key];
                index_1.isAddress(option);
                options.where[key] = option.toLowerCase();
            }
            where += key + ": \"" + options.where[key] + "\"\n";
        }
        var query;
        if (proposalId && !options.where.id) {
            query = graphql_tag_1.default(templateObject_1 || (templateObject_1 = __makeTemplateObject(["query ProposalStakesSearchFromProposal\n        {\n          proposal (id: \"", "\") {\n            id\n            scheme {\n              id\n              name\n            }\n            stakes ", " {\n              ...StakeFields\n            }\n          }\n        }\n        ", "\n      "], ["query ProposalStakesSearchFromProposal\n        {\n          proposal (id: \"", "\") {\n            id\n            scheme {\n              id\n              name\n            }\n            stakes ", " {\n              ...StakeFields\n            }\n          }\n        }\n        ", "\n      "])), proposalId, index_1.createGraphQlQuery(options, where), Stake.fragments.StakeFields);
            return context.getObservableObject(context, query, function (arc, r, queriedId) {
                if (!r) {
                    // no such proposal was found
                    return [];
                }
                var stakes = r.stakes;
                var itemMapper = function (item) {
                    var state = Stake.itemMap(arc, item, queriedId);
                    return new Stake(arc, state);
                };
                if (!stakes) {
                    return [];
                }
                return stakes.map(itemMapper);
            }, (_a = options.where) === null || _a === void 0 ? void 0 : _a.id, apolloQueryOptions);
        }
        else {
            query = graphql_tag_1.default(templateObject_2 || (templateObject_2 = __makeTemplateObject(["query ProposalStakesSearch\n        {\n          proposalStakes ", " {\n              ...StakeFields\n          }\n        }\n        ", "\n      "], ["query ProposalStakesSearch\n        {\n          proposalStakes ", " {\n              ...StakeFields\n          }\n        }\n        ", "\n      "])), index_1.createGraphQlQuery(options, where), Stake.fragments.StakeFields);
            return context.getObservableList(context, query, itemMap, (_b = options.where) === null || _b === void 0 ? void 0 : _b.id, apolloQueryOptions);
        }
    };
    Stake.prototype.state = function (apolloQueryOptions) {
        if (apolloQueryOptions === void 0) { apolloQueryOptions = {}; }
        var query = graphql_tag_1.default(templateObject_3 || (templateObject_3 = __makeTemplateObject(["query StakeState\n      {\n        proposalStake (id: \"", "\") {\n          id\n          createdAt\n          staker\n          proposal {\n            id\n          }\n          outcome\n          amount\n        }\n      }\n    "], ["query StakeState\n      {\n        proposalStake (id: \"", "\") {\n          id\n          createdAt\n          staker\n          proposal {\n            id\n          }\n          outcome\n          amount\n        }\n      }\n    "])), this.id);
        return this.context.getObservableObject(this.context, query, Stake.itemMap, this.id, apolloQueryOptions);
    };
    Stake.fragments = {
        StakeFields: graphql_tag_1.default(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n      fragment StakeFields on ProposalStake {\n        id\n        createdAt\n        dao {\n          id\n        }\n        staker\n        proposal {\n          id\n          scheme {\n            id\n            name\n          }\n        }\n        outcome\n        amount\n      }\n    "], ["\n      fragment StakeFields on ProposalStake {\n        id\n        createdAt\n        dao {\n          id\n        }\n        staker\n        proposal {\n          id\n          scheme {\n            id\n            name\n          }\n        }\n        outcome\n        amount\n      }\n    "])))
    };
    Stake.itemMap = function (context, item, queriedId) {
        if (!item) {
            throw Error("Stake ItemMap failed. " + (queriedId && "Could not find Stake with id '" + queriedId + "'"));
        }
        var outcome = index_1.IProposalOutcome.Pass;
        if (item.outcome === 'Pass') {
            outcome = index_1.IProposalOutcome.Pass;
        }
        else if (item.outcome === 'Fail') {
            outcome = index_1.IProposalOutcome.Fail;
        }
        else {
            throw new Error("Unexpected value for proposalStakes.outcome: " + item.outcome);
        }
        return {
            amount: new bn_js_1.default(item.amount),
            createdAt: item.createdAt,
            id: item.id,
            outcome: outcome,
            proposal: {
                id: item.proposal.id,
                entity: new index_1.Proposals[item.proposal.scheme.name](context, item.proposal.id)
            },
            staker: item.staker
        };
    };
    return Stake;
}(index_1.Entity));
exports.Stake = Stake;
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
